test/fuzz/lua/test_mvcc.lua

```diff
@@ -0,0 +1,872 @@
+    end
+    return key
+end
+
+local function gen_tuple()
+    local tuple = {}
+    for _, field in ipairs(SPACE_FORMAT) do
+        table.insert(tuple, gen_field(field))
+    end
+    return tuple
+end
+
+local function gen_update(fields)
+    return {('{\'=\', 3, %s}'):format(gen_field(fields[1])),
+            ('{\'=\', 4, %s}'):format(gen_field(fields[2]))}
+end
+
+local function gen_rand_operation(ro)
+    local eng = ENGINES[math.random(#ENGINES)]
+    local idx_id = math.random(#eng.idxs) - 1
+    local idx = eng.idxs[idx_id + 1]
+    local op = ro and idx.ro_ops[math.random(#idx.ro_ops)] or
+               idx.ops[math.random(#idx.ops)]
+    local keys = gen_key(idx.fields)
+    if op.type == DQL then
+        if op.subtype == SELECT then
+            local iter = idx.iters[math.random(#idx.iters)]
+            if op.key_cnt == 0 then
+                op.str = op.fmt:format(eng.space, idx_id, iter)
+            elseif op.key_cnt == 1 then
+                op.str = op.fmt:format(eng.space, idx_id, keys[1], iter)
+            elseif op.key_cnt == 2 then
+                op.str = op.fmt:format(eng.space, idx_id, keys[1], keys[2],
+                                       iter)
+            end
+        elseif op.subtype == GET then
+            op.str = op.fmt:format(eng.space, idx_id, keys[1], keys[2])
+        end
+    else
+        local tuple = gen_tuple()
+        local update = gen_update({ SPACE_FORMAT[3], SPACE_FORMAT[4]})
+        if op.subtype == DELETE or op.subtype == UPDATE then
+            op.str = op.fmt:format(eng.space, idx_id, keys[1], keys[2],
+                                   update[1], update[2])
+        else
+            op.str = op.fmt:format(eng.space, tuple[1], tuple[2], tuple[3],
+                                   tuple[4], update[1], update[2])
+        end
+    end
+    return op
+end
+
+-- Choose a random transaction that has not yet been completed.
+local function txs_fetch_incomplete(txs)
+    local tid = math.random(TX_CNT)
+
+    for _ = 1, #txs do
+        if not txs[tid].committed and not txs[tid].aborted then break end
+        tid = tid % #txs + 1
+    end
+
+    if not txs[tid].committed and not txs[tid].aborted then
+        if not txs[tid].running then
+            txs[tid]:begin()
+        end
+    else
+        return
+    end
+    return txs[tid]
+end
+
+local function tx_gen_stmt(tx, ro)
+    tx(gen_rand_operation(ro))
+    if tx.aborted then
+        return
+    end
+
+    local p = math.random()
+    if p < P_ROLLBACK then
+        tx:rollback()
+    elseif p < (P_ROLLBACK + P_COMMIT) then
+        tx:commit()
+    end
+end
+
+local function txs_stop(txs)
+    for tid = 1, #txs do
+        if txs[tid].running then txs[tid]:commit() end
+        table.insert(ro_txs_mask, txs[tid].ro)
+        table.insert(committed_txs_mask, txs[tid].committed)
+        table.insert(bad_dml_txs_mask, txs[tid].bad_dml)
+    end
+end
+
+local function gen_stmts()
```

Comment:
A bit confiusing name - when I see a `gen_*` function, I expect that it only generates it, not executing it as well. Probably, `gen_and_run_stmts` and `tx_gen_and_run_stmt` would be better names, but I'm not sure - they are a bit ugly to me, and I haven't come up with anything better.